rfcs/45-stageless.md

```diff
@@ -0,0 +1,584 @@
+To do so, we need to check both direct *and* transitive strict ordering dependencies, using a depth-first search to walk backwards from the later system using the `CachedOrdering` until a flushing system is reached, and then walking back further from that flushing system until the originating system is reached.
+
+## Drawbacks
+
+1. This will be a massively breaking change to every single Bevy user.
+   1. Any non-trivial control over system ordering will need to be completely reworked.
+   2. Users will typically need to think a bit harder about exactly when they want their gameplay systems to run. In most cases, they should just add the `CoreLabel::AppLogic` label to them.
+2. It will be harder to immediately understand the global structure of Bevy apps.
+   1. Powerful system debugging and visualization tools become even more important.
+3. State transitions are no longer queued up in a stack.
+   1. By default, arbitrarily long chains of state transitions are no longer be processed in the same schedule pass.
+   2. However, users can add as many copies of the `flush_state` system as they would like, and loop it within an exclusive system.
+   3. This also removes "in-stack" and related system groups / logic.
+   4. This can be easily added later, or third-party plugins can create their own abstraction.
+4. It will become harder to reason about exactly when command flushing occurs.
+5. Ambiguity sets are not directly accounted for in the new design.
+   1. They need more thought and are rarely used; we can toss them back on later.
+
+## Rationale and alternatives
+
+### Doesn't label-configuration allow you to configure systems added by third-party plugins?
+
+Yes. This is, in fact, a large part of the point.
+
+The complete inability to configure plugins is the source of [significant user pain](https://github.com/bevyengine/bevy/issues/2160).
+Plugin authors cannot know all of the needs of the end user's app, and so some degree of control must be handed back to the end users.
+
+This control is carefully limited though: systems can only be configured if they are given a public label, and all of the systems under a common label are configured as a group.
+In addition, configuration defined by a private label (or directly on systems) cannot be altered, extended or removed by the app: it cannot be named, and so cannot be altered.
+
+These limitation allows plugin authors to carefully design a public API, ensure critical invariants hold, and make serious internal changes without breaking their dependencies (as long as the public labels and their meanings are relatively stable).
+
+### Why can't you label labels?
+
+If each label is associated with a `SystemConfig`, and part of this struct is which labels are applied, why can't we label labels themselves?
+
+In short: this way lies madness.

```

Comment:
I am alarmingly convinced that madness is the path to take.

Reply:
If labeling is a way of defining a subgraph, then it is very natural to want to create a graph of subgraphs.

*I'll push structured concurrency here again. If we use a schedule as a way of defining a subgraph.  Then creating a graph of those schedules is very easy to visualize. Using system sets gets a little complicated because your sets can be overlapping, and not sub sets of each other.*

Reply:
I would be okay with a schedule being a sequence of subgraphs. However, a graph of graphs is extremely sus to the point I think it's unworkable.

We do not want parallel subgraphs or dependency chains. They can't work because of exclusive systems. It would still be easy to see the *graph*, sure, but it would no longer be easy to interpret which systems can run concurrently.

Even if we could auto-insert flushes, the flushes inserted into a set of parallel subgraphs would basically be random.

Reply:
If all subgraphs have to be atomic forever, then it might be okay. Then if the subgraph had any exclusive systems, it would be ordered strictly before or after other subgraphs.

Reply:
If labels equate to subgraphs, then they might need to be treated as atomic or required to be in a linear sequence.

In any case, it'd be cool if users and plugin authors could exclude certain labels from ordering, to use some as plain tags.

Reply:
> We do not want parallel subgraphs. ...

I think this wouldn't be a huge problem in practice.  We can execute the exclusive systems from both subgraphs in the same phase of the executor. There are probably cases where the user will end up with hard to understand bugs, but I don't think this would be any more problematic than any other case where the user should have ordered their systems, but did not. 

Flushes would be not be random.  They would still only happen when they're required, however we decide that.

Reply:
After everyone weighed in on the Discord thread, I'm convinced that it's fine if this interleaving happens, so long as people have a way out, like having the choice of making their subgraph atomic.

> Flushes would be not be random. They would still only happen when they're required, however we decide that.

The auto-inserted flushes would definitely be correct, it's just that merging subgraphs could put flushes between my systems where I didn't want one. I want to make sure that the user can always decide if interleaving is okay, so there aren't situations were plugins basically kick my systems out of a label (if I don't want them interleaved).